TraderJoe-Bots/Re-entrancy-Transfers/src/agent.ts

```diff
@@ -0,0 +1,80 @@
+import { getEthersProvider } from "forta-agent";
+import { HandleTransaction, TransactionEvent, Finding } from "forta-agent";
+import { createFinding } from "./finding";
+import MarketsFetcher from "./markets.fetcher";
+import NetworkData from "./network";
+import { FUNCTIONS_ABIS, MARKET_UPDATE_ABIS } from "./utils";
+
+let networkManager = new NetworkData();
+let marketsFetcher = new MarketsFetcher(getEthersProvider());
+
+const provideInitialize = (marketsFetcher: MarketsFetcher) => async () => {
+  const { chainId } = await marketsFetcher.provider.getNetwork();
+  networkManager.setNetwork(chainId);
+
+  // set joeTroller contract address
+  marketsFetcher.setJoeTrollerContract(networkManager.joeTroller);
+
+  // fetch Markets from joeTroller
+  await marketsFetcher.getMarkets("latest");
+};
+
+export const provideHandleTransaction =
+  (networkManager: NetworkData, marketsFetcher: MarketsFetcher): HandleTransaction =>
+  async (txEvent: TransactionEvent): Promise<Finding[]> => {
+    const findings: Finding[] = [];
+    // Update the markets list.
+    txEvent.filterLog(MARKET_UPDATE_ABIS, networkManager.joeTroller).forEach((log) => {
+      marketsFetcher.updateMarkets(log.name, log.args.jToken.toString());
+    });
+
+    const monitoredContracts = [
+      Array.from(marketsFetcher.markets).map((market) => market.toLowerCase()),
+      networkManager.sJoeStaking,
+      networkManager.masterChefV2,
+      networkManager.moneyMaker,
+    ].flat();
+
+    // listen to functions calls.
+    const sighashes = txEvent.filterFunction(FUNCTIONS_ABIS, monitoredContracts).map((call) => call.sighash);
+
+    // If any function call on the monitored contracts happened, look for re-entrant calls.
+    if (sighashes.length > 0) {
```

Comment:
Self reviewing.. 

Maybe we can get ride of `filterFunction` usage and just parse manually traces for each transactionEvent. The only optimisation here is avoiding to do that when there's no call, ie `sighashes.length == 0`. 


The problem is that `filterFunction` is already looping over traces, to only decide if there's a call or not. Then we are looping over them again to check for re-entrancy, which is making the bot heavier. 

The alternative solution is removing the selected lines of this comment, and replacing `sighashes` with an array of all monitored functions selectors.  Thoughts on what is more optimized?